File: frontend/src/admin/js/deactivateAllCards.js
import axios from 'axios';
import Swal from 'sweetalert2';
import { domain } from '../../utils';

const deactivateAllCards = async (setCards) => {
    try {
        const { value: password } = await Swal.fire({
            title: 'Enter Admin Password',
            input: 'password',
            inputPlaceholder: 'Password',
            confirmButtonText: 'Login',
            showCancelButton: true,
            inputValidator: (value) => {
                if (!value) {
                    return 'Password is required';
                }
            }
        });
        if (!password) return;
        const response = await axios.post(`${domain}/api/deactivate_all_cards/`, {
            password
        });

        if (response.data.success) {
            Swal.fire({
                icon: 'success',
                title: 'Success!',
                text: response.data.message || 'All activated cards have been deactivated successfully.',
            });
        } else {
            Swal.fire({
                icon: 'info',
                title: 'Info',
                text: response.data.message || 'No Regular Cards to Deactivate.',
            });
        }
        const updatedResponse = await fetch(`${domain}/api/list-cards/`);
        const updatedData = await updatedResponse.json();
        setCards(updatedData);
    } catch (error) {
        console.error('Error:', error);
        if (error.response && error.response.status === 401) {
            Swal.fire({
                icon: 'error',
                title: 'Access Denied',
                text: error.response.data?.message || 'Invalid password.',
            });
        } else if (error.request) {
            Swal.fire({
                icon: 'error',
                title: 'Failed!',
                text: 'Network error. Please check your connection.',
            });
        } else {
            Swal.fire({
                icon: 'error',
                title: 'Failed!',
                text: `Unexpected error: ${error.message}`,
            });
        }
    }
};

export default deactivateAllCards;

File: backend/api/views/cards/deactivate_all_cards.py
from rest_framework import status
from rest_framework.decorators import api_view
from rest_framework.response import Response

from ...models.cards import Card
from ...models.deactivated_cards import DeactivatedCard
from django.utils import timezone
from django.contrib.auth import get_user_model
   

@api_view(['POST'])
def deactivate_all_cards(request):
    password = request.data.get('password')
    if not password:
        return Response({'error': 'Password is required'}, status=status.HTTP_400_BAD_REQUEST)
    user = get_user_model().objects.filter(is_superuser=True).first()
    if user and user.check_password(password):
        cards_to_deactivate = Card.objects.filter(status='Activated', type='Regular')
        
        if cards_to_deactivate.exists():
            now = timezone.localtime(timezone.now())
            cards_to_deactivate.update(status='Deactivated', deactivated_at=now)
            deactivated_cards = [
                DeactivatedCard(
                    card_id=card.card_id,
                    status='Deactivated',
                    deactivated_date=now.date(),
                    deactivated_time=now.strftime('%I:%M %p')
                )
                for card in cards_to_deactivate
            ]
            DeactivatedCard.objects.bulk_create(deactivated_cards)

            return Response({'success': True, 'message': 'All activated cards have been deactivated successfully.'}, status=status.HTTP_200_OK)
        
        return Response({'success': False, 'message': 'No Regular Cards to Deactivate.'}, status=status.HTTP_200_OK)

    return Response({'error': 'Invalid password.'}, status=status.HTTP_401_UNAUTHORIZED)